compatibility/bazel_tools/testing.bzl

```diff
@@ -9,19 +9,37 @@ load("@os_info//:os_info.bzl", "is_linux", "is_windows")
 load("//bazel_tools:versions.bzl", "version_to_name", "versions")
 load("//:versions.bzl", "latest_stable_version")
 
-# Each exclusion in the list above is an object with a range of ledger API test tool versions
-# described by `start` and `end` and a list of platform (ledger) ranges. Each platform range
-# is again described by `start` and `end` as well as the actual list of --exclude flags that
-# should be passed to the ledger API test tool.
-# Start and end are always inclusive and can be omitted if you only need an upper or lower bound.
-# Usually, you'll want to set the latest release from the `LATEST` file to one of the boundaries
-# and use the next version in another one. For example, `1.17.0-snapshot.20210910.7786.0.976ca400`
-# could be the end, while `1.17.0-snapshot.20210910.7786.1` would be the next version.
-# Note that 0.0.0, i.e., current HEAD is considered to come after all other versions.
+# Each exclusion in the list below is defined in terms of:
+#
+# - A range of ledger API test tool versions described by `start` and `end` (both inclusive).
+# - A list of platform (ledger) ranges, each described by `start` and `end` (both inclusive),
+#   as well as the actual list of `--exclude` flags to be passed to the ledger API test tool.
+#
+# Start and end can be omitted and are always inclusive; an interval extreme can be
+# excluded by setting it to a non-existing version that is guaranteed to be between two
+# existing ones, according to version ordering. This is especially useful to denote the
+# upcoming (yet unknown) snapshot version.
+#
+# For example, if the current snapshot is `1.17.0-snapshot.20210811.7565.0.f1a55aa4`, then
+# `1.17.0-snapshot.20210811.7565.1` will be greater than the current snapshot version but
+# smaller than the upcoming snapshot version (because either the date or the commits count
+# or both will increase).
+#
+# An example scenario is a change to both a ledger and the Ledger API Test Tool that makes the
+# latter incompatible with previous versions of the ledger.
+# In this case, you'll want to set the platform `start` and `end` to be unbounded as to cover
+# all existing platform versions (including HEAD) and the test tool `start` to be the upcoming
+# snapshot, which you can denote by using the "intermediate version" trick above.
+# After your change is merged, you'll need a further change to set the platform `end` to the
+# latest released snapshot, so that HEAD and subsequent versions are cross-checked again.
```

Comment:
I don't quite understand this. First off, the goal of these tests is to prevent us from breaking things; exclusions should be used primarily for when we fix bugs in the test tool (in which case it makes sense that _old_ test tools cannot run against _new_ platforms), or when we add features (in which case it makes sense that _new_ tests cannot run against _old_ platforms, where the feature did not exist yet).

In your example, presumably HEAD against HEAD works, or else you shouldn't be able to merge. Therefore, I don't quite get your recommended bounds. It seems to me what you want is to add exclusions for api-test-tool starting at current snapshot +1 and platforms ending at current snapshot +1. That still leaves you running HEAD against HEAD, but, again, presumably that works.

Also I don't see a need to go back and update afterwards, as (again, presumably) newer versions of the test tool and the ledger should work against each other.

Reply:
I'd like to add some common use cases to the doc; you've just enumerated two and the one I refer to is basically their combination (a PR that changes both the ledger API test tool and the platform so that they are compatible with each other but not with previous versions of either).

I realize I am not sure anymore what HEAD is here, though. Is it the result of merging this PR with the target branch at the point the build is started?

Reply:
`HEAD` is "the current commit", i.e. the one the build is being run for. We run the compat tests daily, and for those runs HEAD is "the tip of main at 6AM UTC".

We also run a subset of the compat tests on each PR (IIRC just HEAD against latest release), and in those cases HEAD is the commit being tested by Azure, which is indeed the result of merging the branch with the latest commit on main (at the time of starting that one job, so — at the moment — potentially a different commit per platform).